### PR TITLE
#158320913 Asset Categories List

### PR DESCRIPTION
**Pull request by GbengaOshinaga**

### CODE REVIEW

File: api/views/asset_category.py [hunk 1]
```diff
@@ -30,3 +31,30 @@ def get(self):
...
+    """
+
+    @token_required
+    def get(self):
+        """
+        Gets list of asset categories and the corresponding asset count
+        """
+
+        if request.args:
+            asset_categories = AssetCategory._query(request.args)
```
Comment (GbengaOshinaga) [inline api/views/asset_category.py]: @brayoh I believe we should revisit the implementation of the `_query` method in the `ModelOperations` class, because currently if a `request.args` exists, it doesn't filter the records by which has been deleted or not

Comment (brayoh) [reply api/views/asset_category.py]: Let me have a look then get back to you.

Comment (GbengaOshinaga) [reply api/views/asset_category.py]: @brayoh Alright

Comment (ibutiti) [reply api/views/asset_category.py]: +1 on filtering out deleted records

File: tests/test_asset_categories_endpoints.py [hunk 2]
```diff
@@ -19,3 +20,17 @@ def test_asset_categories_stats_endpoint(self, client, new_asset_category, init_
...
+        response = client.get(f'{api_v1_base_url}/asset-categories',
+                              headers=auth_header)
+        response_json = json.loads(response.data.decode(CHARSET))
+
+        assert response.status_code == 200
+        assert response_json['status'] == 'success'
+        assert isinstance(response_json['data'], list)
+        assert len(response_json['data']) is not 0
+        assert response_json['data'][0]['name'] == 'TestLaptop'
+        asset_category.delete()
```
Comment (ibutiti) [inline tests/test_asset_categories_endpoints.py]: Please add tests for when there are `request.args` including some edge cases like invalid args etc.

File: api/views/asset_category.py [hunk 3]
```diff
@@ -30,3 +31,30 @@ def get(self):
...
+    their corresponding asset counts
+    """
+
+    @token_required
+    def get(self):
+        """
+        Gets list of asset categories and the corresponding asset count
+        """
+
+        if request.args:
```
Comment (ibutiti) [inline api/views/asset_category.py]: You'll need to validate the request args here.

**Comment by brayoh** [comment]
1 small NIT for the README you don't have to create a new branch then pull the code into your local machine you can fetch from the remote into your local machine as follows `git fetch origin ft-asset-categories-list-158320913:ft-asset-categories-list-158320913` then `git checkout ft-asset-categories-list-158320913`

File: api/views/asset_category.py (hunk 3 above)
Comment (brayoh) [reply api/views/asset_category.py]: @GbengaOshinaga Have a look at [ request parsing in flask](https://flask-restful.readthedocs.io/en/0.3.5/reqparse.html) I'd recommend we use this to validate all incoming parameters.

File: api/views/asset_category.py (hunk 1 above)
Comment (brayoh) [reply api/views/asset_category.py]: I'll create  a new ticket for the above

File: tests/test_asset_categories_endpoints.py (hunk 2 above)
Comment (brayoh) [reply tests/test_asset_categories_endpoints.py]: +1 on the above, cover all edge cases

**Comment by brayoh** [comment]
@GbengaOshinaga After you are done with the above changes, remember to `squash` your commits into 1

File: api/views/asset_category.py (hunk 1 above)
Comment (GbengaOshinaga) [reply api/views/asset_category.py]: @brayoh 👍

File: tests/test_asset_categories_endpoints.py (hunk 2 above)
Comment (GbengaOshinaga) [reply tests/test_asset_categories_endpoints.py]: @brayoh @ibutiti Alright, will do

File: api/views/asset_category.py (hunk 3 above)
Comment (GbengaOshinaga) [reply api/views/asset_category.py]: @ibutiti For clarity, what validation do you mean, since validation is already been done on `request.args` in the `DynamicFilter` class?

Comment (ibutiti) [reply api/views/asset_category.py]: oh cool. That's okay.

I see the dynamic filter class validates on the format. I think it would be of value to validate for other things like symbols etc in that filter class. What do you think @brayoh? Another chore?

Comment (brayoh) [reply api/views/asset_category.py]: I'd suggest we have `helper methods` to handle validation of parameter, we write a `decorator function` to handle this, I'll create a chore for this @GbengaOshinaga @ibutiti

**Comment by GbengaOshinaga** [comment]
@brayoh Changes have been implemented

**Comment by brayoh** [comment]
@GbengaOshinaga Kindly fix the merge conflicts

**Comment by GbengaOshinaga** [comment]
@brayoh Merge conflicts have been fixed and tests have passed
